jetty-server/src/main/java/org/eclipse/jetty/server/HttpConnection.java

```diff
@@ -225,36 +228,32 @@ public void onFlushed(long bytes) throws IOException
 
     void releaseRequestBuffer()
     {
-        if (_requestBuffer != null && !_requestBuffer.hasRemaining())
+        if (_retainableByteBuffer != null && !_retainableByteBuffer.hasRemaining())
         {
             if (LOG.isDebugEnabled())
                 LOG.debug("releaseRequestBuffer {}", this);
-            ByteBuffer buffer = _requestBuffer;
-            _requestBuffer = null;
-            _bufferPool.release(buffer);
+            if (_retainableByteBuffer.release() == 0)
+                _retainableByteBuffer = null;
         }
     }
 
     public ByteBuffer getRequestBuffer()
```

Comment:
You broke already some APIs, I would make this `private` or even remove the method and inline its 2 usages.

Reply:
Done.